src/lock.rs

```diff
@@ -153,49 +118,94 @@ impl Lock {
                 name: name.to_string(),
             });
         }
-        Ok((changed, res))
+        Ok(changed)
+    }
+
+    pub fn freeze(&mut self, deps: &[FrozenDependency]) {
+        self.patch_existing_deps(&deps);
+        self.add_missing_deps(&deps);
+    }
+
+    fn add_missing_deps(&mut self, frozen_deps: &[FrozenDependency]) {
+        let known_names: &Vec<_> = &mut self.dependencies.iter().map(|d| d.name()).collect();
+        let new_deps: Vec<_> = frozen_deps
+            .iter()
+            .filter(|x| !known_names.contains(&&x.name))
+            .collect();
+        for dep in new_deps {
+            let locked_dep = SimpleDependency::from_frozen(dep);
+            println!("+ {}", locked_dep.line);
+            self.dependencies.push(LockedDependency::Simple(locked_dep));
+        }
+    }
+
+    fn patch_existing_deps(&mut self, frozen_deps: &[FrozenDependency]) {
+        for dep in &mut self.dependencies {
+            match dep {
+                LockedDependency::Git(_) => (),
+                LockedDependency::Simple(s) => {
+                    Self::patch_existing_dep(s, frozen_deps);
+                }
+            }
+        }
+    }
+
+    fn patch_existing_dep(
+        dep: &mut SimpleDependency,
+        frozen_deps: &[FrozenDependency],
+    ) -> Option<()> {
+        let frozen_match = frozen_deps.iter().find(|x| x.name == dep.name);
+        let frozen_match = frozen_match?;
```

Comment:
Did you mean to put the try on its own line?
No big deal if you did, at least it's very explicit